### ISSUE TITLE
Single calc instance and window vs. Multiple calc windows

### ISSUE THREAD

**Issue by TristanYoung** [Microsoft/calculator#215]
I own many Microsoft keyboards with calculator launch buttons.  Very handy button.  I use it some 100x a day (minimum)

Every time I press the calculator button, a new calculator window opens (multi-window).

I would like to see a single-window mode, where repeated presses of the calculator button brings an already-running instance to the front with input focus (single-window single-instance).

Entering numbers from a variety of sources, and being able to work with a single calculator would be the most efficient for some people. ALT-TABing to an already-running calculator is less efficient.  But I can also see the value in having multi-windows, and some people prefer that behaviour.

Respecting choice, single-window / multi-window modes should be user-configurable.

**Comment by mdtauk** [comment]
Originally it was a single instance app, but the feedback wanted it to support multi-instance.  I agree it could be added as an optional setting.  Software like VLC gives the choice

**Comment by rudyhuyn** [comment]
What if we add a command line parameter like `--singleInstance` to prevent the creation of extra windows, you will then be able to modify `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\AppKey\18` and run `calc.exe --singleInstance` instead of `calc.exe`.

Would it be ok for you?

**Comment by TristanYoung** [comment]
Thank you for suggesting the command-line parameter.  Adding a command-line parameter would be a great idea.

Will there be opportunity to advertise/expose this functionality to others somehow - like in the help documents?

**Comment by TristanYoung** [comment]
I still have XP machines, and the default XP calculator app is supposed to be single instance I believe.  It behaves that way, except when OpenOffice Calc opens up - for some reason multiple calculators open up whenever OOC is running.  Close OOC, and then the calculator app returns to normal.  Heheheh.

**Comment by MicrosoftIssueBot** [comment]
This is your friendly Microsoft Issue Bot. I've seen this issue come in and have gone to tell a human about it.

**Comment by rudyhuyn** [comment]
I've a diff ready, but there is some little tricks I want to confirm with the community and Microsoft before sending it.

We can't use `calc.exe` in command lines, it looks like Windows uses a "fake" .exe stored somewhere on Windows to launch the custom protocol (`ms-calculator` or `calculator` used by the win10 calculator).

In fact, this project doesn't even use the recently added UWP CommandLine handler (Windows 10 1803). So to be able to use command line, we can't use `calc.exe`, we need another name. I was thinking about `Calculator.exe` but please, send suggestions!

So let me confirm if:
- we want the application to manage a `--singleInstance` command line argument
- keep calc.exe working (of course)
- add the UWP AppExecutionAlias `Calculator.exe` with the optional `--singleInstance` parameter.

**Comment by TristanYoung** [comment]
Thank your for working on this idea/behaviour.

Personally, I don't mind if it's done via command-line, or even a little push-pin icon to set single-instance or multi-instance.

Creating a separate Calculator.exe will satisfy the needs, however it's too bad that you have to create a new name and can't reuse the existing one.

There are times when someone might want to switch to multi-instance, after doing a bunch of calculations in the single instance calculator, and then switch back.

Are we trying to keep the GUI simple and uncluttered, and is it feasible to make it switchable after it's already launched, and storing the single/multi instance state?  I also like the idea of providing a visible toggle for less-savvy users, and would get around having to create the second executable.

Just throwing that out there.  I'm grateful for any improvements :)

**Comment by rudyhuyn** [comment]
I don't think to toggle will be that useful. You already have 2 ways to launch the calculator and already have the choice between:
- click on the icon in the App list (start menu) -> multi instance
- click on the icon in the task bar -> will open the existing window (or create a new one if first instance)

The big issues to fix are:
- the calculator key on Keyboards
- allow scripts (Dos, Powershell...) to select how they want to launch the app.

For both scenarios, the command line parameter can fix the issue.

**Comment by GeeLaw** [comment]
Why not just make `ms-calculator:` protocol aware of the strings passed to it? See [App.xaml.cpp L457](https://github.com/Microsoft/calculator/blob/0f615faff93dc11101f6b730ce46c69f5390b94c/src/Calculator/App.xaml.cpp#L457).

PowerShell can just run `Start-Process ms-calculator:lastusedview|scientific`.

**Comment by rudyhuyn** [comment]
We can also modify the protocol and add a parameter, but the command line will still be necessary for all applications expecting (and only supporting) a path and not a protocol. I'm not sure AppKeys for example support protocols

**Comment by grochocki** [comment]
Thanks for the feature suggestion. This does sound like an interesting idea. We have a [process in place](https://github.com/Microsoft/calculator/blob/master/docs/NewFeatureProcess.md) for feature suggestions that impact the end-user experience, which includes a template that helps frame the feature pitch.  Please update using that template.

> We can't use `calc.exe` in command lines, it looks like Windows uses a "fake" .exe stored somewhere on Windows to launch the custom protocol (`ms-calculator` or `calculator` used by the win10 calculator).

We rely on the `calc.exe` in System32 for backwards compatibility reasons (e.g., references to "C:\Windows\System32\calc.exe")

![image](https://user-images.githubusercontent.com/1414423/54100200-8e101300-437a-11e9-84b4-1c1bb174bc45.png)

**Comment by grochocki** [comment]
@TristanYoung When you get the chance, can you update your idea with more details following the feature pitch template from our [New Feedback Process](https://github.com/Microsoft/calculator/blob/master/docs/NewFeatureProcess.md)?

**Comment by TristanYoung** [comment]
I'm used to fixing problems head-on, rather than discussing them.  I'll give it a shot though.

**Problem Statement**
Entering calculations into the calculator can be tedious when working across multiple program windows. Compounding the problem is when multiple calculator instances are running. Selecting the wrong instance can cause calculation errors, and wastes time.  This can be detrimental when calculating prices and margins, precise dimensions, etc.  Some tasks are better left to spreadsheets, but for quick calculations, a calculator is optimal.

Many keyboards have a calculator button for quick access to the calculator. This makes calling up a calculator instance quick and easy the first time, but for subsequent recall of the same instance, the button is useless as it launches additional instances.

Create a calculator instance that can launched by hotkey or calculator button, that can be easily recalled without ALT-TABBING or WIN-TABBING through 20 or so windows. The target audience is anyone who requires a single instance and speed. It's common to have a single handheld calculator, rather than a desk full of calculators - as it simplifies working at the desk - apply that to the Windows desktop.

A multi-instance-only calculator, or a single-instance-only calculator is not universally appealing.

My own uses are mixed business/financial, machining and optical calculations, and home/school.  The goal is to improve speed and accuracy, reduce error and frustration.


**Evidence or User Insights**
I don't have so much any evidence or user insight, other than my own use-cases.  I have google searched and found  many people complaining about multiple calculator instances, and trying to figure out ways to force single instance behaviour.

Single-instance is the way the calculator used to operate. It's no different than having a single handheld calculator.  There will be people that want single-instance, and other people that will want multiple-instance, and even some that will want to switch between the two on-demand.  It's reasonable to believe that not everyone likes things the same cookie-cutter way.  That is one of the things that made Windows go from a horrible user experience to a great one - has lead to Windows 10's success over Windows 8.

 
**Proposal**
- State-savable toggle button  (1st choice - most visible, most easy)
- State-savable menu option (2nd choice - less visible, awkward)
- Command-line parameter (worst choice - invisible to less-technically-inclined users)

Being able to change the single/multi instance state in an easy manner should meet everyone's needs.

If a button is used (my strong preference):

The wording or glyph representing the option should impart the meaning - such as:
	- Single Window / Multi Window
	- Dynamic glyph using the existing calculator icon:
		- Single icon = single instance
		- Two overlapping icons = multiple instance

When single-instance state is enabled in one instance, and there are other instances running, the user should be asked if they are sure they want to close the other instances. The default should be NO, in order to prevent the loss of data in other instances.

The question should be answerable using the user's input method of choice:
	- the mouse to click on relatively-positioned YES / NO buttons that don't require exteme mouse movements to reach (suitable for ultra-wide display users)
	- TAB to change between NO / YES, ENTER to select the choice
	
	
**Goals**
- Adapt the calculator so that it functions the way some people expect or need it to.
- Provide greater customization based on the sometimes dynamic needs of the user.
- Reduce the time it takes to activate the one specific calculator instance needed.
- Reduce error caused by entering calculations into wrong instances.
 
 
**Non-Goals**
Cater only to advanced users familiar with command-line parameters.


**Low-Fidelity Concept**
It's a toggle button for single instance/multi instance, that also saves the state to the registry, and can close multiple instances. A picture can be supplied if it's really necessary.

Sorry if this isn't what you had in mind.

**Comment by grochocki** [comment]
Thanks, @TristanYoung!

**Comment by TristanYoung** [comment]
You're welcome Grochocki.  Sorry, I guess I should have edited my original post.

**Comment by mdtauk** [comment]
VLC offers a choice on how multiple-instancing works.
![image](https://user-images.githubusercontent.com/7389110/55827796-f2053300-5b02-11e9-83ce-95a1821de825.png)

You could have a toggle button.

**Allow a single instance of Calculator [ ||| ]____]**

and have the default set to Off.

**Comment by rudyhuyn** [comment]
Good job @tristanYoung, I fully agree with you

One important piece of information missing here is: why is the calculator multi-instance? While I understand that some users need a multi-instance app, I believe the majority of users would prefer only one instance.

I see 2 possible reasons:
- User studies show that a majority of users prefer multi-instance, in software development, we should never assume how users use your app, we can be surprised sometimes.
- the Windows team needed a first party application to demo UWP multi-instances, but this choice wasn't backed by data or studies.

If it's because of reason 2, it should be reconsidered and a user study can help determine the app's default behavior.

I'm in favor of 2 checkboxes, similar to VLC (thank you Martin for the screenshot, happy to see you here!):
- allow multi instance
- use only one instance when started from keyboard shortcut (only enabled when the first checkbox is checked).

But should the first one check by default? We can't decide without data.

**Comment by mdtauk** [comment]
Having one calculator open in programmer mode, and another in standard mode can be very useful.

But having the option would make the most sense.

**Comment by TristanYoung** [comment]
rudyhuyn, from what I understand, people wanted multi-instance, and so it was changed. Calculator for XP was single instance. I can't remember, but I think Windows 7 Calculator was operating as single-instance for me, but it depended on the Microsoft keyboard driver version I used.  I also found OpenOffice Calc was one application that would cause the singe-instance feature in XP to break and would not only launch multiple instances, would cause two windows (OO-Calc + Windows Calculator) to be focused at the same time (figure that one out :), which caused me performance problems.

I agree with mdtauk, someone might need multiple calculators configured in different calculator styles.

A brief survey of people at work plus others elsewhere, some say a single instance is their preference, while others think a multi-instance would work best for them. The people who cited using ALT-TAB to cycle to the calculator rarely have more than 2 or 3 windows open, and don't understand the trials of ALT-TABBING through 20 or so windows.

Since I wear a gazillion hats at work, and sometimes perform my multiple roles near-simultaneously, I speak from the point-of-view of the sales manager, shipper, developer of our company website, book keeper and bill payer.  I know for me, a single calculator is greatly appreciated, for speed and accuracy.

**Comment by TristanYoung** [comment]
> But should the first one check by default? We can't decide without data.

Doing a user study seems to be complicating things.  Since it's multi by default now, why not just keep it multi by default, and let users make the choice to change it?

If a user study is a requirement, let's get it done.  Never started one before - how should we start?

**Comment by grochocki** [comment]
Thanks everyone for the conversation on this issue. Generally speaking, we like the current multi-instance behavior in cases where the user explicitly launches another instance of Calculator (e.g., middle clicking the task bar icon). However, we recognize there are cases where you may favor programmatically opening an existing instance, when possible (i.e., hardware calculator keyboard button). As such, we would like to move forward with the command parameter/protocol approach, recognizing that there would likely be some work that a Microsoft employee will need to do to update calc.exe.

There has also been some conversation about making this an in-app setting. We considered this approach when reviewing the pitch, and while we certainly _could_ go this route, we generally try to avoid adding "advanced" settings that might be confusing to typical users. I am sure many folks in this thread would not consider this to be an advanced setting, but we have to consider how users of all backgrounds and skill sets use Calculator. Ultimately, we felt that the command line/protocol approach provides enough flexibility to address the pain point without introducing too much complexity.

**Comment by rudyhuyn** [comment]
have a very old PR ready adding a command parameter and forcing single instance. Once specs done (including the parameter name) I will update it and submit it.

**Comment by alystair** [comment]
Any update re: this? 

In the interim I created a AutoHotKey script that overrides current behaviour. It focuses currently instance and launches if no instance is running. https://github.com/alystair/my-productivity-stack/blob/master/Launch%20Calculator.ahk

**Comment by mikewizc** [comment]
Any update?
After install Powertoys, everytime I wanna open calc, it opens a new instance...

**Comment by henriksen** [comment]
Yeah, that's actually a new annoyance. @alystair 's AHK script fixed the calc button on my keyboard, but now when I'm using PowerToys Run to start it on my laptop (with no HW key), I get a new instance everytime. Having a command line switch isn't that useful there. 

Maybe there could be some way of setting a default externally to the calculator app? Perhaps a registry key or a settings file in the user directory?

**Comment by grochocki** [comment]
> I have a very old PR adding a command parameter and forcing single instance ready. Once specs done (including the parameter name) I will update it and submit it.

@rudyhuyn Thoughts on what the parameter name should be?

Once #596 is checked in, we can explore adding a setting like this:

![image](https://user-images.githubusercontent.com/1414423/89863386-d975b480-db5e-11ea-9ebf-7befe256867f.png)

I am still concerned that typical users might find this type of setting confusing, so I am curious if we should hide this setting (and I suspect others in the future) behind an "Advanced options" navigation or toggle, like this:

![image](https://user-images.githubusercontent.com/1414423/89863401-e1cdef80-db5e-11ea-80c8-23a8ba9a857d.png)

![image](https://user-images.githubusercontent.com/1414423/89863640-5e60ce00-db5f-11ea-9593-50970c4ab80c.png)

**Comment by henriksen** [comment]
I think hiding it behind an "Advanced settings" link is quite reasonable and in line with what I've seen in for instance Windows Settings.

**Comment by Jaiganeshkumaran** [comment]
> I think hiding it behind an "Advanced settings" link is quite reasonable and in line with what I've seen in for instance Windows Settings.

Yes but full screen pages should not be used unless there's a lot of options

**Comment by horze21** [comment]
hi guys, was this feature added to stop multiple instances... i have been looking for the above settings posted in photo... but cant seem to locate it?
thanks in advanced

**Comment by grochocki** [comment]
This feature has not yet been implemented and is dependent on #596, which is still open.

**Comment by outdooracorn** [comment]
Can I request that this is implemented as "Allow only one instance **per virtual desktop**"? When I open the calculator app in Desktop 3, I don't want to be taken back to Desktop 1 and shown the calculator app which is already open there.
Thanks

**Comment by alystair** [comment]
> I am still concerned that typical users might find this type of setting confusing, so I am curious if we should hide this setting (and I suspect others in the future) behind an "Advanced options" navigation or toggle

@grochocki 'typical' users wouldn't even open the settings to begin with. I feel it would be pretty safe to not overcomplicate things. KISS :)

**Comment by fuweichin** [comment]
As an UWP app, Calculator is designed to work in desktop / laptop / tablet, so instead of adding a command-line arggument, adding a setting like "Allow only one instance" seems to be a better choice, since the calc.exe is now changed to play a role of shortcut (like .lnk shortcut).

I use the `Calc` key on keyboard to launch Calculator, multiple calculator instances cause some mess. Before the issue gets refined, I would rather press `F12` key to use browser Console as a calculator.


By the way: 
I'm trying to make a single instance demo but get stucked.

+ [x] ensure single instance
+ [ ] activate previous instance
  + [ ] find process of previous instance (a very complicated method for UWP app process is found)
  + [ ] activate main window of the process (a method for UWP app process is not found yet)

The methods below is working on Win32 apps but not on UWP apps.

```
        [DllImport("user32.dll")]
        private static extern bool ShowWindow(IntPtr hWnd, int nCmdShow);
        // Sets window to be foreground
        [DllImport("user32.dll")]
        private static extern int SetForegroundWindow(IntPtr hwnd);
```

Anybody knows how to activate an UWP app window by process info?

**Comment by Jaiganeshkumaran** [comment]
@[fuweichin](https://github.com/fuweichin) UWP apps are already single-instance by default but can opt-in for multi-instancing. I believe Calculator actually uses a multi-window approach instead of multiple instances (so all of them are in the same process).

**Comment by jordiorlando** [comment]
With 4 years having passed since the initial feature request, I assume that this is dead in the water. Hopefully someone will take it upon themselves to open a PR - unfortunately I don't possess the skills to do so myself. I would however happily toss some cash to whomever gets this across the line!

**Comment by codesimon** [comment]
An "Allow only one instance" setting would solve so many problems.

**Comment by hschlossberg** [comment]
How difficult does this need to be to either add a parameter or simple user 'Settings' option?!?!  Having these multiple calc windows is so very frustrating! In Windows 10, the old calc.exe app could still be run in compatibility mode.  But in Win 11, we are stuck with this "modern" app that has this critical "feature" missing!

**Comment by 5haun** [comment]
This article might help someone willing to take on trying to enable single instance for the calculator app: [https://blogs.windows.com/windowsdeveloper/2022/01/28/making-the-app-single-instanced-part-3/](https://blogs.windows.com/windowsdeveloper/2022/01/28/making-the-app-single-instanced-part-3/)

**Comment by dcoffin88** [comment]
If anyone is still looking for this feature, changing SupportsMultipleInstances to 0 in the registry works. I'm sure this will have to be changed again after any app updates. It still opens multiple instances if you launch the Calculator from the start menu, but the keyboard calc button and Win+R(calc) are limited to single instance with this change.

Here is the reg key that I found and changed on my instance

`[HKEY_CLASSES_ROOT\Extensions\ContractId\Windows.Protocol\PackageId\Microsoft.WindowsCalculator_11.2508.1.0_x64__8wekyb3d8bbwe\ActivatableClassId\App.AppXc11dqpdjvk55gd6z0w7qmyvrsxymf1tk.mca\CustomProperties]
"Name"="ms-calculator"
"SupportsMultipleInstances"=dword:00000000`

**Comment by jayjar291** [comment]
Modifying the registry did not work. Multiple instances still launch. A single-instance calculator is sorely needed for Windows.
